Add integration tests for Job Push

berkaycanbc opened the pull request:
## Description

<!-- Please explain the changes you made here. -->

IT tests for the Job Push is missing. To cover the happy path scenario end to end, we need to implement IT tests. These cases should also cover closing a stream which will actually test the onClose hook. Please refer to these two discussions for more details:
https://github.com/camunda/zeebe/pull/13351#discussion_r1263414875
https://github.com/camunda/zeebe/pull/13351#discussion_r1263834385

## Related issues

<!-- Which issues are closed by this PR or are related -->

closes #13516 

<!-- Cut-off marker
_All lines under and including the cut-off marker will be removed from the merge commit message_

## Definition of Ready

Please check the items that apply, before requesting a review.

You can find more details about these items in our wiki page about [Pull Requests and Code Reviews](https://github.com/camunda/zeebe/wiki/Pull-Requests-and-Code-Reviews).

* [ ] I've reviewed my own code
* [ ] I've written a clear changelist description
* [ ] I've narrowly scoped my changes
* [ ] I've separated structural from behavioural changes
-->

## Definition of Done

<!-- Please check the items that apply, before merging or (if possible) before requesting a review. -->

_Not all items need to be done depending on the issue and the pull request._

Code changes:
* [ ] The changes are backwards compatibility with previous versions
* [ ] If it fixes a bug then PRs are created to [backport](https://github.com/camunda/zeebe/compare/stable/0.24...main?expand=1&template=backport_template.md&title=[Backport%200.24]) the fix to the last two minor versions. You can trigger a backport by assigning labels (e.g. `backport stable/1.3`) to the PR, in case that fails you need to create backports manually.

Testing:
* [ ] There are unit/integration tests that verify all acceptance criterias of the issue
* [ ] New tests are written to ensure backwards compatibility with further versions
* [ ] The behavior is tested manually
* [ ] The change has been verified by a QA run
* [ ] The impact of the changes is verified by a benchmark

Documentation:
* [ ] The documentation is updated (e.g. BPMN reference, configuration, examples, get-started guides, etc.)
* [ ] If the PR changes how BPMN processes are validated (e.g. support new BPMN element) then the Camunda modeling team should be informed to adjust the BPMN linting.

Other teams:
If the change impacts another team an issue has been created for this team, explaining what they need to do to support this change.
- [ ] [Operate](https://github.com/camunda/operate/issues)
- [ ] [Tasklist](https://github.com/camunda/tasklist/issues)
- [ ] [Web Modeler](https://github.com/camunda/web-modeler/issues)
- [ ] [Desktop Modeler](https://github.com/camunda/camunda-modeler/issues)
- [ ] [Optimize](https://github.com/camunda/camunda-optimize/issues)

Please refer to our [review guidelines](https://github.com/camunda/zeebe/wiki/Pull-Requests-and-Code-Reviews#code-review-guidelines).

berkaycanbc committed:
feat(qa): add integration tests for Job Push

Add happy path integration test for Job Push

npepinpe commented on qa/integration-tests/src/test/java/io/camunda/zeebe/it/clustering/JobPushIT.java:
```diff
@@ -0,0 +1,115 @@
+
+      // when
+      final List<ActivatedJob> jobs = new ArrayList<>();
+      client
+          .newStreamJobsCommand()
+          .jobType("foo")
+          .consumer(jobs::add)
+          .timeout(Duration.ofMinutes(2))
+          .send();
+
+      client
+          .newCreateInstanceCommand()
+          .processDefinitionKey(deploymentEvent.getProcesses().get(0).getProcessDefinitionKey())
+          .requestTimeout(Duration.ofMinutes(1))
+          .send()
+          .join();
```
:x: This section is flaky for two reasons:

1. We can only push jobs to a stream if the stream existed before the job handled (i.e. created, timed out, failed, etc. - anything that would make it activate-able again).
2. Opening a stream is asynchronous, so there's a chance the process instance is created, and the job is created, before the stream is opened.

There are some options to work around this here, but your best bet is to wait for my PR with actuators. Then you can actually query the broker to ensure the stream exists before creating the process instance.

> **Note**
> In production, since users will be using a job worker, this is not a big deal since we're both streaming and polling.

npepinpe replied:
As this may block you, we can do the other changes for this PR, and then I'm happy to take adapt your PR once mine is merged so you can already do other stuff.

berkaycanbc replied:
Alright, these makes sense. Then, let's wait for the blocking PR to be merged. Afterwards, please feel free to add changes.

npepinpe commented on qa/integration-tests/src/test/java/io/camunda/zeebe/it/clustering/JobPushIT.java:
```diff
@@ -0,0 +1,115 @@
+
+      // then
+      // wait for the record to be available and consumed
+      engine.waitForIdleState(Duration.ofSeconds(30));
+      assertThat(records())
+          .as("job should become activated immediately")
+          .extracting(Record::getValueType, Record::getIntent)
+          .containsSubsequence(Tuple.tuple(ValueType.JOB_BATCH, JobBatchIntent.ACTIVATED));
+      assertThat(jobs).hasSize(1);
```
:x: It's possible for the job batch activated record to be written, but the job to not have been sent or received by the stream consumer yet. Instead, I think it's perfectly fine to just use Awaitility and wait until we receive the expected job within 10 seconds or so. That means we probably don't need the whole ContainerEngine here.

A more complete alternative would be to wait for the job, complete it, and then assert that the process instance is completed. But I don't think it's necessary here.

berkaycanbc replied:
We'll wait #13684 to be merged to properly write assertions to the IT test.

Cc: @npepinpe

npepinpe reviewed (changes requested):
Thanks for the PR @berkaycanbc!

:x: Unfortunately, it's flaky in two places. I highlighted them in the comments.

:question: I also had one question about the use of containers here. I'm not sure we really need it for this kind of test. It would probably be fine to use the `ClusteringRuleExtension` or something like that.

berkaycanbc commented on qa/integration-tests/src/test/java/io/camunda/zeebe/it/clustering/JobPushIT.java:
```diff
@@ -0,0 +1,115 @@
+@Testcontainers
+final class JobPushIT {
+  private final Network network = Network.newNetwork();
+  private final ZeebeCluster cluster =
+      ZeebeCluster.builder()
+          .withBrokersCount(1)
+          .withGatewaysCount(1)
+          .withPartitionsCount(1)
+          .withEmbeddedGateway(false)
+          .withImage(ZeebeTestContainerDefaults.defaultTestImage())
+          .build();
```
I think, using an embedded gateway should just work since we are testing the behaviour here and it does not require that. 👍

berkaycanbc committed:
refactor(qa): add embedded gateway for Job Push integration tests

The container setup is converted in a way that it will have only one broker container
and an embedded gateway along with it.
